diff.py

```diff
@@ -2148,9 +2148,12 @@ class ArchSettings:
 
 SH2_SETTINGS = ArchSettings(
     name="sh2",
-    re_int=re.compile(r"[0-9]+"),
+    # don't hexify anything since mov instructions are not prefixed with 0x
+    # so they get double-converted (i.e 44 -> 0x2c when it should be 0x44)
+    # any immediate values other than the branch instructions are prefixed with #
+    re_int=re.compile(r""),
```

Comment:
Here's the mov.l example that prompted the hexify change. https://gist.github.com/sozud/15de54c4149b67f5d55f0b30d6761361 This is the objdump output before it goes into Display(), printed from a local install of decomp.me so it should be accurate to how asm-differ is used in that context. func_060E8780 is the assembled "theirs" code and .text() is the C version.

The issue is, for example that this line gets matched:
```
  16:   d1 0b           mov.l   44 <func_060E8780+0x44>,r1      ! 60e87d0
```
44 looks like an integer to that regex but it's actually hex without a prefix. I should probably fix the relocs so that it points to a label but figured I should try and fix the hexify issue first. I think the regex could probably be improved to handle this, I can see about cataloging the instructions that work this way and developing a better regex. I'm not super familiar with sh2 assembly but the way the manual describes this instruction looks more like this `/* 0xD10B */ mov.l @(0x02E, pc), r1`, so objdump is using a shorthand that I'm not sure that gnu assembler can reassemble or not. Just to be clear the instruction is moving the data relative to the program counter into r1 according to the offset.

For immediate instructions, the objdump output looks like this:
```
        # 0:   ec 01           mov     #1,r12
        # 2:   71 01           add     #1,r1
        # 4:   ec ff           mov     #-1,r12
        # 6:   71 ff           add     #-1,r1
        # 8:   ec 7f           mov     #127,r12
        # a:   71 7f           add     #127,r1
        # c:   ec 80           mov     #-128,r12
        # e:   71 80           add     #-128,r1
```
So it's signed 8 bit values prefixed with #.

I'll see if I can improve the regex and try `instructions_with_address_immediates`.

Reply:
I updated the re_int regex and added a test for branch instructions. I'm not sure how to get mov instructions marked with ~> arrows.

Reply:
Interesting!

> so objdump is using a shorthand that I'm not sure that gnu assembler can reassemble or not

Yeah, that's common. Same usually goes for branches, on all arches (relative in the instr encoding but made absolute by objdump, but lacking the sort of target labels you'd want for real assemblable output).

instructions_with_address_immediates currently assumes that the imm is the last instruction argument, which isn't the case for your mov; fixing that may involve some work. Upside is that it would make diffs relative:
```
00: mov.l 04,r1
02: whatever
04: .word 0x12345678
```
vs
```
00: mov.l 02,r1
02: .word 0x12345678
```
would only show a diff for `whatever` being deleted, not the mov argument differing, because the targets resolve to the same thing.

Arm32 has a very similar thing to this, with specific "load from (pc-relative) pool" instructions, which we handle by parsing it specifically and ignoring the address (iirc), instead parsing out the comment at the end of line that objdump adds that shows the loaded value. Not totally happy about how that mechanism is implemented but could be worth taking inspiration from still, if you feel like continuing to hack on this.